### PR TITLE
Implement StreamTable and StreamTableProvider (#7994)

### LINKED ISSUE #7994: Decouple Streaming Use-Case from File IO Abstractions
### Is your feature request related to a problem or challenge?

Currently we accommodate streaming workloads within DataFusion by overloading the file IO abstractions. 

This is not always a very good fit and results in a number of workarounds:

* Providing PartitionedFile information to methods that write new files via FileSinkConfig - https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/struct.FileSinkConfig.html#structfield.file_groups
* Passing unbounded_input to parallel write logic that really shouldn't care - https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/src/datasource/file_format/write/orchestration.rs#L79
* Special logic to handle reading files that aren't regular files - https://github.com/apache/arrow-datafusion/pull/7282#discussion_r1294019448
* The somewhat confusing notion of an "infinite" file - https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/struct.FileScanConfig.html#structfield.infinite_source
* An ObjectStore::append method that doesn't naturally fit with the design goals of the project - https://github.com/apache/arrow-rs/pull/4978/files#diff-fbf0cdd67af2b24703a831d28c932bfe0535234e3b745d3a5437b9c0bc881ddeR89
* Schema inference within ListingTable currently assumes files are finite
* ObjectMeta is passed around in plans with the implicit assumption that files are not changing underneath
* And more...

As DataFusion gets more sophisticated about handling catalogs, reading/writing partitioned data, caching data, this overloading is getting more and more arcane and hard to reason about, and I think it is overdue we do something to address it.

### Describe the solution you'd like

I would like to separate the notions of FileSink and FileScan from a StreamSink and StreamSource, this would allow abstractions that better fit their respective use-cases.

In particular 

* FileSink and FileScan can focus on reading/writing partitioned immutable files following standard big data practices
* StreamSink and StreamSource can focus on reading/writing CSV / JSON (/ Avro) data from streaming sources

Not only would this simplify the current code, but would also expand the streaming support in DataFusion

* Allows for more efficient non-blocking IO, as linux FIFO's support poll(2) (unlike general files)
* Potential future integrations with data streaming systems such as Kafka, etc...

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tustvold**
## Which issue does this PR close?

Closes #7994

## Rationale for this change

See ticket

## What changes are included in this PR?

Implements StreamTable and StreamTableProvider as in the [design proposal](https://github.com/apache/arrow-datafusion/issues/7994#issuecomment-1806791702). Also defines a DefaultTableFactory that wires this up as the default for unbounded tables.

### CODE REVIEW

File: datafusion/core/tests/fifo.rs [hunk 1]
```diff
@@ -172,48 +308,27 @@ mod unix_test {
...
-        // Timeout for a long period of BrokenPipe error
-        let broken_pipe_timeout = Duration::from_secs(10);
-        // Spawn a new thread to write to the FIFO file
-        thread::spawn(move || {
-            let file = OpenOptions::new().write(true).open(file_path).unwrap();
-            // Reference time to use when deciding to fail the test
-            let execution_start = Instant::now();
-            write_to_fifo(&file, &header, execution_start, broken_pipe_timeout).unwrap();
-            for (cnt, line) in enumerate(lines) {
-                while waiting_lock.load(Ordering::SeqCst) && cnt > wait_until {
```
Comment (tustvold) [inline datafusion/core/tests/fifo.rs:191]: I had some difficulty following the various locking, etc... I think the new implementation is broadly equivalent, but I may be missing some nuance

File: datafusion/core/tests/fifo.rs [hunk 2]
```diff
@@ -17,32 +17,208 @@
...
+        }
+
+        async fn write_all(
+            &self,
+            mut data: SendableRecordBatchStream,
+            _context: &Arc<TaskContext>,
+        ) -> Result<u64> {
+            let config = self.0.clone();
+            let (sender, mut receiver) = tokio::sync::mpsc::channel::<RecordBatch>(2);
+            // Note: FIFO Files support poll so this could use AsyncFd
```
Comment (tustvold) [inline datafusion/core/tests/fifo.rs:188]: This is an example of the potential benefits of a custom TableProvider, it can use external knowledge about the nature of the files in question.

File: datafusion/core/tests/fifo.rs [hunk 3]
```diff
@@ -394,8 +491,6 @@ mod unix_test {
                 a2  INT NOT NULL
             )
             STORED AS CSV
-            WITH HEADER ROW
```
Comment (tustvold) [inline datafusion/core/tests/fifo.rs:397]: There is no particular reason that CSV headers could not be supported, but it was simpler to just not read/write them for the purposes of this test

**Comment by metesynnada** [comment]
Thx @tustvold, I believe that the proof of concept is looking good. My opinion:

Instead of creating a new TableProvider for FIFO, we could use a `StreamingTableProvider` including its factory. We can obtain the correct factory by using the UNBOUNDED (can be change like STREAM) keyword in SQL. Additionally, we can leverage a new `StreamSinkExec` and valid `DataSink` implementations to handle different use cases like FIFO or Kafka. I think this approach could scale effectively. For FIFO, there are different types of formats as well, also a potential Kafka Sink may implement different types of `BatchSerializer`s. This was in my mind for a time but couldn't find the necessary time to code it. Does it seem like fine plan?

**Comment by metesynnada** [comment]
I believe we can use your implementation of DataSink for FIFO, nicely coded 👍🏻

**Comment by tustvold** [comment]
> Instead of creating a new TableProvider for FIFO, we could use a StreamingTableProvider including its factory

So this is what I started on implementing, but I wasn't really sure how this interface would materially differ from the existing interfaces used by this PR. In particular I'm struggling to devise an interface that would substantially reduce the amount of boilerplate, whilst still preserving the necessary flexibility. Perhaps I am missing something?

**Review by alamb — COMMENTED** [review]
I reviewed this PR and I think it looks really nice -- thank you @tustvold  and @metesynnada  🏆 

I also filed https://github.com/apache/arrow-datafusion/pull/8035 with some additonal information about `PartitionStream`

**Comment by metesynnada** [comment]
> > Instead of creating a new TableProvider for FIFO, we could use a StreamingTableProvider including its factory
> 
> So this is what I started on implementing, but I wasn't really sure how this interface would materially differ from the existing interfaces used by this PR. In particular I'm struggling to devise an interface that would substantially reduce the amount of boilerplate, whilst still preserving the necessary flexibility. Perhaps I am missing something?

To help with this subject, let's break it down into smaller parts. We can begin by focusing on the FIFO table and finding ways to optimize the SQL involved. Additionally, for external tables that are unbounded, we can create FIFO tables with the proper formatting, starting with CSV and AVRO. When it comes to session state, we keep `TableProviderFactory` as

```rust
    /// This is used to create [`TableProvider`] instances for the
    /// `CREATE EXTERNAL TABLE ... STORED AS <FORMAT>` for custom file
    /// formats other than those built into DataFusion
    table_factories: HashMap<String, Arc<dyn TableProviderFactory>>,
``` 
which is not scalable. Currently, Datafusion SQL is closely intertwined with the ListingTable approach, which resulted in FIFO support within the ListingTable. To avoid this, we need to be able to use multiple `TableProviderFactory` instances for a single format, by specifying the connector.

TL:DR, this part should change to scale the support.

```rust
async fn create_custom_table(
        &self,
        cmd: &CreateExternalTable,
    ) -> Result<Arc<dyn TableProvider>> {
        let state = self.state.read().clone();
        let file_type = cmd.file_type.to_uppercase();
        let factory =
            &state
                .table_factories
                .get(file_type.as_str())
                .ok_or_else(|| {
                    DataFusionError::Execution(format!(
                        "Unable to find factory for {}",
                        cmd.file_type
                    ))
                })?;
        let table = (*factory).create(&state, cmd).await?;
        Ok(table)
    }
```

**Comment by tustvold** [comment]
> which is not scalable

I don't see why you couldn't have a TableProviderFactory that itself delegates out to other TableProviderFactory based on some criteria? Perhaps I am misunderstanding what you mean by scalable?

**Comment by tustvold** [comment]
I've marked this ready for review as I think the feedback is this is a sound approach, but feel free to correct me if I am wrong

**Review by ozankabak — COMMENTED** [review]
Please do not merge this yet -- AFAICT this breaks use cases including creating external tables reading from FIFO files that used to have built-in support.

We like the idea of having a `FIFOTable`, separating a single sink into `FileSink` and a `StreamSink` etc. However, we don't seem to agree on the design yet. We will soon share our design proposal with diagrams and API level specs.

**Comment by tustvold** [comment]
> AFAICT this breaks use cases including creating external tables reading from FIFO files that used to have built-in support

TBC this PR simply updates the fifo tests and makes no changes to ListingTable, which would instead be a follow on PR. The idea of this PR is to show that removing the functionality from ListingTable, whilst it would still represent a breaking change, would not require major architectural changes to existing workloads, nor would it require introducing new abstractions to DF. If you feel this hasn't been demonstrated please let me know how I can articulate this better.

**Comment by ozankabak** [comment]
We have discussed this today in detail. We will come back with a design/proposal early next week. Let's iterate together to find the best design. We will also help with the implementation when we converge.

**Comment by tustvold** [comment]
Given I think this PR has demonstrated the existing APIs can accommodate streaming use-cases, perhaps we might move forward with this and potentially refine things later? I'm keen to find a way to keep things moving so we don't hold up #8029 and by extension the next DF release

**Comment by alamb** [comment]
I don't think we should merge this without getting consensus / ensuring the synnada team has reasonable way forward, especially there are other options (e.g. an arrow patch release) that can unblock the datafusion release

In particular, I think the tradeoffs between different implementation options and order come down to who is going to do the work, and I think there are ways it doesn't all fall into @tustvold . I will try and set up a call / some forum to discuss this and will post notes on the outcome

**Comment by ozankabak** [comment]
We anticipate to have @metesynnada's design proposal, which aims to embody the discussions in Issue #7994, in a weeks time. The proposed design will provide an API for streaming sources that will decouple them from `ListingTable`. We propose to not merge this draft, iterate on the upcoming proposal to converge on a design as a community, and hold off on upgrading to arrow-rs 49.0.0 during this process. We do so as this upgrade removes the append functionality from object_store, which is critical to the current implementation of FIFO functionality in Datafusion.

We aim to resolve the situation in approximately two weeks, with the following end state: (1) We will have merged the converged-on design decoupling ListingTable from streaming sources, (2) We will proceed with the upgrade to arrow-rs 49.0.0.

In the meantime, we may make a serde-fix-backported 48.1.0 release of arrow-rs to get Datafusion benefit from this fix.

**Comment by alamb** [comment]
> We anticipate to have @metesynnada's design proposal, which aims to embody the discussions in Issue #7994, in a weeks time. The proposed design will provide an API for streaming sources that will decouple them from `ListingTable`. We propose to not merge this draft, iterate on the upcoming proposal to converge on a design as a community, and hold off on upgrading to arrow-rs 49.0.0 during this process. We do so as this upgrade removes the append functionality from object_store, which is critical to the current implementation of FIFO functionality in Datafusion.
> 
> We aim to resolve the situation in approximately two weeks, with the following end state: (1) We will have merged the converged-on design decoupling ListingTable from streaming sources, (2) We will proceed with the upgrade to arrow-rs 49.0.0.
> 
> In the meantime, we may make a serde-fix-backported 48.1.0 release of arrow-rs to get Datafusion benefit from this fix.

I think this makes a lot of sense -- thank you @ozankabak . I will prepare a 48.0.1 patch release candidate -- see https://github.com/apache/arrow-rs/issues/5050#issuecomment-1802574764

**Comment by alamb** [comment]
Marking as draft so we don't accidentally merge this as we work on the updated design

**Comment by alamb** [comment]
> We have discussed this today in detail. We will come back with a design/proposal early next week. Let's iterate together to find the best design. We will also help with the implementation when we converge.

@ozankabak  -- @tustvold said he may have some time to help (and would be interested in assisting the design and implementation to help accelerate this process).

Is it possible to describe the use case you have in more detail? 

Given that this particular PR replicates the functionality that already exists in DataFusion using the extension APIs, we think we must not fully understand the use case for which you are designing a solution for.

**Comment by ozankabak** [comment]
I will talk to @metesynnada tomorrow and check with him to see where he is at in terms of the proposal. If he is close to getting something he can share soon, it'd be better to just talk about the proposal and collaborate on it directly. If he is not close to having something publishable, we will explain the use cases we are trying to support beforehand. I will circle back tomorrow.

> @tustvold said he may have some time to help (and would be interested in assisting the design and implementation to help accelerate this process)

thank you, we appreciate it

**Comment by alamb** [comment]
It seems to me one of the core tensions here is that `ListingTable` (the built in table source from DataFusion) uses `ObjectStore`to read data. 

Therefore, if `append` is removed from `ObjectStore` ListingTable can no longer be used directly to append to existing files (e.g. FIFO), though it can make new files in a directory. 


Thus one API question  is probably  "should ListingTable" support the ability to modify / append existing files, or should that functionality be delegated to some other structure, possibly even outside the core DataFusion crate.

**Comment by tustvold** [comment]
> Thus one API question is probably "should ListingTable" support the ability to modify / append existing files

The proposal in https://github.com/apache/arrow-datafusion/issues/7994 would posit no

> uses ObjectStore to read data.

I think it is more subtle than that, the very notion of how ListingTable works is predicated on the notion of immutable files. If files can change following planning, etc... you aren't going to get coherent query results.

**Comment by alamb** [comment]
>> Thus one API question is probably "should ListingTable" support the ability to modify / append existing files
> The proposal in https://github.com/apache/arrow-datafusion/issues/7994 would posit no

Right, but we have at least one set of users who use that functionality, so I don't think we can just unilaterally remove functionality without getting consensus that moving the functionality downstream makes sense

**Comment by tustvold** [comment]
> we can just unilaterally remove functionality without getting consensus that moving the functionality downstream makes sense

I don't think anyone is proposing unilaterally removing anything, the whole exercise has been trying to find a path to accommodate these workloads whilst fixing ListingTable.

**Comment by metesynnada** [comment]
The problem is that this PR becomes "ready to merge" instead of a POC PR. I got the idea and iterated on this to come up with a design, as promised. I will publish it today. Indeed, there is enough functionality for FIFO or more use cases without the object store. I agree with separating batch and stream sources to ease the feature development on both sides.

However, pushing this PR means deleting the gatekeepers of the unbounded functionality without an alternative. I do not understand the rush to disable these tests from the rest of the code base. What is the gain here, can you explain if I do not fully understand?

**Comment by metesynnada** [comment]
I updated an initial design document on https://github.com/apache/arrow-datafusion/issues/7994#issuecomment-1806791702

**Comment by tustvold** [comment]
I've updated this inline with the design articulated by @metesynnada in - https://github.com/apache/arrow-datafusion/issues/7994#issuecomment-1806791702

As StreamingTable already exists, I had to opt for StreamTable instead

**Review by metesynnada — COMMENTED** [review]
Thank you for making progress on this. With a few changes, I think we can use this PR as a stopgap and a springboard to accelerate the implementation of the [proposal](https://github.com/apache/arrow-datafusion/issues/7994#issuecomment-1806791702).  The key change required for this to be a stopgap is preserving support for `CREATE UNBOUNDED EXTERNAL TABLE`.

Consider the following test code:
```rust
let mut state = SessionState::new_with_config_rt(config, runtime);
let mut factories = HashMap::with_capacity(1);
factories.insert(
“CSV”.to_string(),
Arc::new(StreamTableFactory::default()) as _,
);
*state.table_factories_mut() = factories;
let ctx = SessionContext::new_with_state(state);
```

In order to support  `CREATE UNBOUNDED EXTERNAL TABLE` functionality, we need to consider both the format and the source type (unbounded/streaming or bounded/file) in factory logic.

Currently, we are using the format as sole the key for table factories, 

https://github.com/apache/arrow-datafusion/blob/aaaf698b5e46dd1c4d29bce69dc539a667d2dda6/datafusion/core/src/execution/context/mod.rs#L1275-L1293)

which hinders our ability to distinguish between factories for the same format.

A possible solution would be to refine the key to include both the format and its boundedness. Since `create_custom_table`

https://github.com/apache/arrow-datafusion/blob/aaaf698b5e46dd1c4d29bce69dc539a667d2dda6/datafusion/core/src/execution/context/mod.rs#L738-L756 

already provides information about unboundedness, it could be utilized to accurately determine the appropriate `TableFactory`.

We would be happy to sign off on this as a stopgap if we can get `CREATE UNBOUNDED EXTERNAL TABLE` working. We would then base our work on top of this to get the proposal implemented more quickly than otherwise.

**Comment by tustvold** [comment]
Is the desired behaviour that StreamTable would only be used for unbounded tables, and otherwise ListingTable used?

> We would then base our work on top of this to get the proposal implemented more quickly than otherwise.

Thank you for being flexible, this situation is definitely not ideal

**Comment by metesynnada** [comment]
> Is the desired behaviour that StreamTable would only be used for unbounded tables, and otherwise ListingTable used?
> 
> >

Yes, this would be enough for separation.

**Comment by tustvold** [comment]
Sorry for the delay was pulled into a meeting, https://github.com/apache/arrow-datafusion/pull/8021/commits/0e0ff1484d031c2d47eb5e37aa32e671ed3cf02a adds a `DefaultTableFactory` that automatically switches between `ListingTableFactory` and `StreamTableFactory` based on the unbounded parameter. This I _think_ provides the desired behaviour out of the box behaviour, whilst allowing people to explicitly opt-in to using a specific variant.

**Comment by metesynnada** [comment]
I believe this behavior is satisfactory.

File: datafusion/core/src/datasource/mod.rs [hunk 4]
```diff
@@ -58,29 +57,33 @@ fn create_ordering(
...
+                        Ok(expr) => {
+                            sort_exprs.push(PhysicalSortExpr {
+                                expr,
+                                options: SortOptions {
+                                    descending: !sort.asc,
+                                    nulls_first: sort.nulls_first,
+                                },
+                            });
+                        }
+                        // Cannot find expression in the projected_schema, stop iterating
```
Comment (tustvold) [inline datafusion/core/src/datasource/mod.rs:75]: This is necessary as the sort key may reference columns not found in the projected schema.

This change matches get_projected_output_ordering

File: datafusion/core/src/datasource/stream.rs [hunk 5]
```diff
@@ -177,7 +190,7 @@ impl TableProvider for StreamTable {
     }
 
     fn table_type(&self) -> TableType {
-        TableType::Temporary
+        TableType::Base
```
Comment (tustvold) [inline datafusion/core/src/datasource/stream.rs:193]: This is necessary so that Drop behaves correctly

**Comment by tustvold** [comment]
unbounded_file_with_symmetric_join has failed on OS X, I have no looked into if this is an intermittent failure, but I note that this test has been ignored since https://github.com/apache/arrow-datafusion/commit/d6c2233e21b6aa7cfb157e5f36c75934f077c167#diff-2f6b05fb3bc4e54422c680dfa47dd8a993cc93fcaa056f678e455cc5d83ae6d0R237

I'm not sure what the history here is, is this a known flaky test that I should keep ignored?

**Comment by ozankabak** [comment]
Thanks @tustvold, we will look into this early morning tomorrow and report our findings

**Comment by metesynnada** [comment]
The previous locks ensure that the FIFO is read and written in a streaming manner, simulating infinite data streams. This guarantees that each batch produced contains a specific amount of data.

I plan to submit a corrective commit soon to address stucking caused by asynchronous processing and configuration issues in the current implementation.

**Comment by tustvold** [comment]
Thank you, I've also just pushed a commit that configures the readers to respect the configured batch size which may help

**Comment by metesynnada** [comment]
> Thank you, I've also just pushed a commit that configures the readers to respect the configured batch size which may help

Yes, it is necessary as well. However, I need to figure out some points on async runtime. Then, we can move on I think.

**Comment by tustvold** [comment]
> However, I need to figure out some points on async runtime

If you're able to articulate the issue I may be able to help out here.

**Comment by tustvold** [comment]
CI failures are due to logical merge conflicts - https://github.com/apache/arrow-datafusion/pull/8187

File: datafusion/core/src/datasource/stream.rs [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+impl PartitionStream for StreamRead {
+    fn schema(&self) -> &SchemaRef {
+        &self.0.schema
+    }
+
+    fn execute(&self, ctx: Arc<TaskContext>) -> SendableRecordBatchStream {
+        let config = self.0.clone();
+        let schema = self.0.schema.clone();
+        let batch_size = ctx.session_config().batch_size();
+        let mut builder = RecordBatchReceiverStreamBuilder::new(schema, 2);
```
Comment (metesynnada) [inline datafusion/core/src/datasource/stream.rs:258]: I am attempting to eliminate the tokio channel dependency, which is not necessary in theory. However, the `dyn RecordBatchReader<Item = Result<arrow_array::RecordBatch, arrow_schema::ArrowError>>` is not implementing Send and Sync. Therefore, creating an unfold directly from them is not possible. Do you have any suggestions on how to achieve this without creating a dedicated stream struct similar to what we are doing in ExecutionPlan.execute()?

Comment (tustvold) [reply datafusion/core/src/datasource/stream.rs:258]: The tokio channel is necessary because we're performing blocking file IO no?

Comment (metesynnada) [reply datafusion/core/src/datasource/stream.rs:266]: For synchronous IO, using `spawn_blocking` is the rule of thumb, you are right. We can keep this.

Comment (tustvold) [reply datafusion/core/src/datasource/stream.rs:266]: Yeah, it is actually a bug that we don't appear to currently be doing this. I've filed https://github.com/apache/arrow-datafusion/issues/8188 which I'll follow up with once this is merged in

**Review by metesynnada — APPROVED** [review]
I think we can merge this when the tests are passed.

**Comment by tustvold** [comment]
Thank you for your help getting this over the line.

@ozankabak any final comments or are you happy for me to proceed?

**Comment by ozankabak** [comment]
@tustvold, thank you for helping out with the stopgap. @metesynnada will use this as a foundation as he implements the proposal to completion. Feel free to go ahead and merge this after CI passes 🚀

File: datafusion/core/src/datasource/stream.rs [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+            .with_header(cmd.has_header)
+            .with_batch_size(state.config().batch_size());
+
+        Ok(Arc::new(StreamTable(Arc::new(config))))
+    }
+}
+
+/// The data encoding for [`StreamTable`]
+#[derive(Debug, Clone)]
+pub enum StreamEncoding {
```
Comment (alamb) [inline datafusion/core/src/datasource/stream.rs:75]: 👍

File: datafusion/sqllogictest/test_files/ddl.slt [hunk 8]
```diff
@@ -750,7 +750,7 @@ query TT
 explain select c1 from t;
 ----
 logical_plan TableScan: t projection=[c1]
-physical_plan CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/empty.csv]]}, projection=[c1], infinite_source=true, has_header=true
+physical_plan StreamingTableExec: partition_sizes=1, projection=[c1], infinite_source=true
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/ddl.slt:753]: 👍

File: datafusion/core/src/datasource/stream.rs [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+            "csv" => Ok(Self::Csv),
+            "json" => Ok(Self::Json),
+            _ => plan_err!("Unrecognised StreamEncoding {}", s),
+        }
+    }
+}
+
+/// The configuration for a [`StreamTable`]
+#[derive(Debug)]
+pub struct StreamConfig {
```
Comment (alamb) [inline datafusion/core/src/datasource/stream.rs:96]: Should we note here that this is likely to undergo API changes to support additional usecases (like streaming to/from Kafka)?

File: datafusion/core/src/datasource/stream.rs [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            }
+            StreamEncoding::Json => {
+                let file = OpenOptions::new().write(true).open(&self.location)?;
+                Ok(Box::new(arrow::json::LineDelimitedWriter::new(file)))
+            }
+        }
+    }
+}
+
+/// A [`TableProvider`] for a stream source, such as a FIFO file
```
Comment (alamb) [inline datafusion/core/src/datasource/stream.rs:183]: Can we maybe also add some comments here about how this is similar / different than `ListingTable` to help new people navigate the code and understand when to use one vs the other?

We could totally do this as a follow on PR as well

**Review by alamb — APPROVED** [review]
Thank you everyone -- I know there were some communication bumps getting here, but this looks like a really nice design in my mind and sets us up for good things to come.

Thanks again

File: datafusion/core/src/datasource/stream.rs (hunk 10 above)
Comment (tustvold) [reply datafusion/core/src/datasource/stream.rs:183]: I will add some comments in a follow on PR

File: datafusion/core/src/datasource/stream.rs (hunk 9 above)
Comment (tustvold) [reply datafusion/core/src/datasource/stream.rs:96]: I'm not sure this would necessarily change materially, it parallels ListingTableConfig which is used by ListingTable, but I suppose we shall see

File: datafusion/core/src/datasource/stream.rs (hunk 10 above)
Comment (alamb) [reply datafusion/core/src/datasource/stream.rs:183]: follow up: https://github.com/apache/arrow-datafusion/pull/8192
